src/Propulsion.SqlStreamStore/LockedIngester.fs

```diff
@@ -0,0 +1,229 @@
+                            let mutable submitted = false
+                            try
+                                // Obtained the lock, check batch against committed position.
+                                let! position = ledger.GetPosition { Stream = streamId; ConsumerGroup = consumerGroup }
+
+                                stats.UpdateReadPosition(position)
+
+                                match InternalBatch.sliceAt position batch with
+                                | Ahead batch ->
+                                    logger.Debug(
+                                        "Batch at position {firstPosition} through {lastPosition} is ahead of committed position {committedPosition}, processing.",
+                                        batch.firstPosition, batch.lastPosition, position)
+                                    do! submit (batch.firstPosition, commit handle batch, batch.messages) |> Async.Ignore
+                                    submitted <- true
+                                | Overlapped (batch, firstPosition) ->
+                                    logger.Debug(
+                                        "Batch at position {firstPosition} through {lastPosition} is overlapping committed position {committedPosition}, processing from committed position.",
+                                        firstPosition, batch.lastPosition, position)
+                                    do! submit (batch.firstPosition, commit handle batch, batch.messages) |> Async.Ignore
+                                    submitted <- true
+                                | Behind ->
+                                    logger.Debug(
+                                        "Batch at position {firstPosition} through {lastPosition} is behind committed position {committedPosition}, skipping.",
+                                        batch.firstPosition, batch.lastPosition, position)
+                            finally
+                                if not submitted then
+                                    // No processing to be done on this batch, release the lock.
+                                    logger.Debug("Releasing the lock without submission")
+                                    do handle.Dispose()
+
+                            // Release the current batch to pick a new batch on the next iteration.
+                            workItem <- Work.Take
+                        else
+                            // Failed obtaining the lock, retry with the same work item.
+                            logger.Debug("Failed to obtain the lock")
+                    with
+                    | exc ->
+                        logger.Error(exc, "Exception while running StreamReader loop")
+                        return! Async.Raise exc
+            }
+
+        interface IDisposable with
+            member this.Dispose() =
+                work.CompleteAdding()
+                work.Dispose()
```

Comment:
Not really feeling strong on how a blocking collection should be disposed in a clean way - any suggestions?